src/MongoDB.Driver.Core/Core/Compression/Zstd/ZstandardStream.cs

```diff
@@ -0,0 +1,382 @@
+        private readonly StreamWriteHelper _streamWriteHelper;
+        private readonly CompressionMode _compressionMode;
+
+        private bool _disposed;
+        private bool _flushed;
+        private readonly ZstandardNativeWrapper _nativeWrapper;
+
+        public ZstandardStream(
+            Stream compressedStream,
+            CompressionMode compressionMode,
+            Optional<int> compressionLevel = default)
+        {
+            _compressedStream = Ensure.IsNotNull(compressedStream, nameof(compressedStream));
+            _compressionMode = EnsureCompressionModeIsValid(compressionMode);
+
+            _nativeWrapper = new ZstandardNativeWrapper(_compressionMode, EnsureCompressionLevelIsValid(compressionLevel));
+            switch (_compressionMode)
+            {
+                case CompressionMode.Compress:
+                    _streamWriteHelper = new StreamWriteHelper(
+                        compressedStream: compressedStream,
+                        compressedBuffer: _arrayPool.Rent(_nativeWrapper.RecommendedOutputSize));
+                    break;
+                case CompressionMode.Decompress:
+                    _streamReadHelper = new StreamReadHelper(
+                        compressedStream: compressedStream,
+                        compressedBuffer: _arrayPool.Rent(_nativeWrapper.RecommendedInputSize));
+                    break;
+            }
+        }
+
+        public override bool CanRead => _compressedStream.CanRead && _compressionMode == CompressionMode.Decompress;
+
+        public override bool CanWrite => _compressedStream.CanWrite && _compressionMode == CompressionMode.Compress;
+
+        public override bool CanSeek => false;
+
+        public override long Length => throw new NotSupportedException();
+
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+
+            if (!_disposed)
+            {
+                switch (_compressionMode)
+                {
+                    case CompressionMode.Compress:
+                        try
+                        {
+                            if (!_flushed)
+                            {
+                                try
+                                {
+                                    Flush();
+                                }
+                                catch
+                                {
+                                    // ignore exception
+                                }
+                            }
+                        }
+                        finally
+                        {
+                            _nativeWrapper.Dispose();
+                            _arrayPool.Return(_streamWriteHelper.CompressedBufferInfo.Bytes);
+                        }
```

Comment:
I thought we had discussed that finally is not appropriate here. Finally is typically used to do cleanup when there is a failure, not as a way to ignore exceptions. Lines 119-137 should be:

    if (!_flushed)
    {
        try { Flush(); } catch { }
    }
    _nativeWrapper.Dispose();
    _arrayPool.Return(_streamWriteHelper.CompressedBufferInfo.Bytes);